### PR TITLE
Add GORM v2 compatibility

### PR DESCRIPTION
**Pull request by dmalch-hashicorp**
Updated portions of the codebase to be compatible with a newer version of GORM. This PR is similar to https://github.com/hashicorp/go-gin-gorm-opencensus/pull/10, but takes it further adding full support of OpenCensus reporting. Tested in the resource manager repo.

Additionally, the following changes have been made:
- The callbacks were refactored to be compatible with GORM v2. The registration process was updated to handle and propagate errors, and the internal callback logic was updated to match GORM v2 API changes.
- Removed an unused spanScopeKey and refactored the handling of the trace context to use idiomatic context usage in Go for tracing, replacing the need to pass around the gorm DB instance.
- The statistics measurement definitions and views were centralized in the original `ocgorm` package. This simplified the stats.go file and promoted the reuse of metrics.

I've divided the changes into individual commits to ensure a better and more effective review experience, with a focus on each commit individually.

### CODE REVIEW

File: pkg/ocgormv2/callbacks.go
Comment (byronwolfman) [inline pkg/ocgormv2/callbacks.go:188]: Is this a metric, and if so, are we tagging the metric with `time.Now()`?

Or is this a trace tag?

**Comment by manish-hashicorp** [comment]
Cool! Out of curiosity did you try adding the changes in existing `pkg/ocgorm` package to make it compatible with v1 and v2?

File: pkg/ocgormv2/callbacks.go
Comment (dmalch-hashicorp) [reply pkg/ocgormv2/callbacks.go:188]: Yes, It is used to record the start time of a database operation.  Later, this start time is retrieved from the context, and If the start time exists, it is parsed back into a `time.Time` value. The elapsed time of the operation is then calculated by subtracting this start time from the current time. This elapsed time is recorded as a metric, specifically as the latency of the operation in milliseconds. 

This code is not new,  it already existed in a previous version of the library utilized within the cloud-sdk. This addition is only to support for Gorm v2.

**Comment by dmalch-hashicorp** [comment]
> Cool! Out of curiosity did you try adding the changes in existing `pkg/ocgorm` package to make it compatible with v1 and v2?

I tried, but unfortunately, the signatures in Gorm v2 have changed, which required the creation of new functions. Tiago's branch inspired the decision to put it in a new package. However, it's still possible to store everything in the same package as v1 and rename the new functions. For example, instead of using `ocgormv2.RegisterCallbacks`, we could use `ocgorm.RegisterCallbacksV2`.

**Comment by manish-hashicorp** [comment]
> However, it's still possible to store everything in the same package as v1 and rename the new functions.

Nah. I think its fine to put it in another package. I am just wondering if there is some duplicate code that we can be avoided. As in if a fix is done in one it will need to be done in another.

File: pkg/ocgormv2/callbacks.go
Comment (byronwolfman) [reply pkg/ocgormv2/callbacks.go:188]: Thanks! I think this originally spooked me because it looked like it was part of a _metric_, but if I'm reading this right, it's part of a _trace_ span.

**Comment by dmalch-hashicorp** [comment]
JFYI, I have successfully tested resource-manager with Gorm v2 and this library in dev environment. All metrics and tracing functions worked as anticipated ([DD link](https://hashicorp-cloud.datadoghq.com/apm/services/cloud-resource-manager.db/operations/opencensus/resources?dependencyMap=qson%3A%28data%3A%28dependencyLayout%3AhybridView%2CisAccordionOpen%3A%21f%29%2Cversion%3A%210%29&env=dev&fromUser=false&graphType=flamegraph&groupMapByOperation=null&isInferred=false&panels=qson%3A%28data%3A%28%29%2Cversion%3A%210%29&resources=qson%3A%28data%3A%28visible%3A%21t%2Chits%3A%28selected%3Atotal%29%2Cerrors%3A%28selected%3Atotal%29%2Clatency%3A%28selected%3Ap95%29%2CtopN%3A%2110%29%2Cversion%3A%211%29&shouldShowLegend=true&sort=time&spanID=7399408730761118461&spanViewType=metadata&summary=qson%3A%28data%3A%28visible%3A%21t%2Cerrors%3A%28selected%3Acount%29%2Chits%3A%28selected%3Aversion_rate%29%2Clatency%3A%28selected%3Alatency_by_version%2Cslot%3A%28agg%3A99%29%2Cdistribution%3A%28isLogScale%3A%21f%29%29%2Csublayer%3A%28slot%3A%28layers%3Aservice%29%2Cselected%3Aabsolute%29%29%2Cversion%3A%211%29&topGraphs=latency%3Alatency%2Chits%3Aversion_count%2Cerrors%3Aversion_count%2CbreakdownAs%3Aavg_time&traceQuery=&version=679001-8f00240&view=spans&start=1710838073120&end=1711038550000&paused=true) to the service page with this change active)

**Review by manish-hashicorp — APPROVED** [review]
LGTM from me. Thanks a lot for doing this! Will be great to get approval from SRE team if possible.

**Review by tscolari — APPROVED** [review]
This looks great!

**Review by byronwolfman — APPROVED** [review]

**Review by morganebridges — APPROVED** [review]

File: pkg/ocgormv2/callbacks.go
Comment (Dombo) [inline pkg/ocgormv2/callbacks.go:1]: Probably the most annoying review comment to receive post-merge, is it possible to have a single package with various methods for V1 and V2 gorm? With the approach of copying the code to swap out a few parts of it I'm concerned what'll happen over time, will the two diverge unintentionally for instance.

It's probably we'll have gormv1 and v2 co-existing for years to come FWIW

Comment (dmalch-hashicorp) [reply pkg/ocgormv2/callbacks.go:1]: Thank you for looking into it, and as I replied in another comment https://github.com/hashicorp/go-gin-gorm-opencensus/pull/11#issuecomment-2013162564, yes it is possible, we could store everything in the same package as v1 and rename the new functions. For example, we could rename `ocgormv2.RegisterCallbacks` to `ocgorm.RegisterCallbacksV2`. However, this approach would not allow us to reuse the code, since the function signatures have changed considerably in v2, so it's just about moving the same new functions into the old package.

Comment (Dombo) [reply pkg/ocgormv2/callbacks.go:1]: Can you link me a diff of how the function signatures would change? It's hard to spot any changes cause everything in this new package is ++ lines.

Comment (dmalch-hashicorp) [reply pkg/ocgormv2/callbacks.go:1]: The difficulties with reusing the same signatures are caused by changes like:
* Constructions like `db.Callback().Create().Before("gorm:create").Register("instrumentation:before_create", c.beforeCreate)` now return an error
* Callbacks now have a signature `func(db *gorm.DB, operation string)` instead of `func(scope *gorm.Scope, operation string)`
* The way how the current table is retrieved has changed from `scope.TableName()` to `db.Statement.Table`
* A similar change for the current SQL script `scope.SQL` -> `db.Statement.SQL.String()`
* The new Gorm v2 supports Go contexts, eliminating the need to store contexts in the scope field as done in Gorm v1

I would say the commit https://github.com/hashicorp/go-gin-gorm-opencensus/pull/11/commits/ffb6081e6a2a7dbab0d41f03b742356e3e5fa29f provides the best overview of the changes required, it is based on the file copied from v1 and contains only the changes to support v2.

Comment (Dombo) [reply pkg/ocgormv2/callbacks.go:1]: Thanks for putting the effort into evaluating this review comment! I checked out the branch you built and [compared against main](https://github.com/hashicorp/go-gin-gorm-opencensus/compare/c8291d61e529640413224828394ee75267f140c2...30e5d8156471ca7430a3c6627e423312ee5fe21f) before this changeset merged.

The surface area to keep in sync between the approach we've taken here vs the one I was enquiring about is effectively the same. I don't think it's worth the version bump. I guess this dependency is unlikely to change often and if it breaks it only breaks a portion of observability. I'm okay with leaving as is unless you feel strongly otherwise.

Comment (dmalch-hashicorp) [reply pkg/ocgormv2/callbacks.go:1]: Thank you for reviewing it. I am looking forward to using it in our codebase. 👍

**Pull request merged**
